### PR TITLE
Better synchronisation between Gutenberg and Core code

### PR DESCRIPTION
**Pull request by youknowriad**
In the ideal scenario, Gutenberg's php code should be composed of three things:

 - A `lib/compat` folder with all the changes that are in Core or supposed to land in the next version in Core. (Most of the php code)
 - PHP code that get updated and back ported on each release automatically. For now, we have just two things here: Dynamic blocks php code and block supports.
 - Glue code that is specific to the plugin: "demo page for instance..."

Right now, we still have a lot of php code that is "unclear": lives outside of `lib/compat` but is already on core or its status need to be clarified. This kind of code makes it very hard to maintain the plugin and do the backports on each WordPress release. Ideally, we should seek to remove that code or move it to the right place. This PR starts that work by focusing on two FSE files.

### CODE REVIEW

File: lib/compat/wordpress-5.9/theme-templates.php [hunk 1]
```diff
@@ -68,4 +68,7 @@ function gutenberg_filter_wp_template_unique_post_slug( $override_slug, $slug, $
 
 	return $override_slug;
 }
+
+// Remove 5.8 filter if existant.
+remove_filter( 'pre_wp_unique_post_slug', 'wp_filter_wp_template_unique_post_slug' );
```
Comment (youknowriad) [inline lib/compat/wordpress-5.9/theme-templates.php:73]: That filter exists on 5.8 but is slightly different.

File: lib/full-site-editing/edit-site-page.php [hunk 2]
```diff
@@ -152,7 +152,7 @@ static function( $classes ) {
 		'siteUrl'                              => site_url(),
 		'postsPerPage'                         => get_option( 'posts_per_page' ),
 		'styles'                               => gutenberg_get_editor_styles(),
-		'defaultTemplateTypes'                 => gutenberg_get_indexed_default_template_types(),
+		'defaultTemplateTypes'                 => get_default_block_template_types(),
```
Comment (youknowriad) [inline lib/full-site-editing/edit-site-page.php]: Gutenberg plugin is using a special function here that is different from Core, I was not entirely sure which one was the right one, but I decided to use the Core approach. Let me know if this is wrong.

File: lib/full-site-editing/template-loader.php [hunk 3]
```diff
@@ -12,8 +12,8 @@ function gutenberg_add_template_loader_filters() {
 	if ( ! gutenberg_supports_block_templates() ) {
 		return;
 	}
-
-	foreach ( gutenberg_get_template_type_slugs() as $template_type ) {
+	$template_type_slugs = array_keys( get_default_block_template_types() );
+	foreach ( $template_type_slugs as $template_type ) {
 		if ( 'embed' === $template_type ) { // Skip 'embed' for now because it is not a regular template type.
 			continue;
 		}
```
Comment (youknowriad) [inline lib/full-site-editing/template-loader.php:19]: The template loader is something that landed on 5.8 I believe but was tweaked on 5.9, so ideally the code here should be moved somehow to `lib/compat/5.9` folder maybe. cc @ockham

**Comment by talldan** [comment]
> A lib/compat folder with all the changes that are in Core or supposed to land in the next version in Core. (Most of the php code)

To be honest, I didn't know this until recently. I wonder if other contributors might be the same. It might be good to put a README in the folder to explain how it works. I don't mind having a look at doing that (hopefully I have the right understanding now).

**Comment by youknowriad** [comment]
> To be honest, I didn't know this until recently. I wonder if other contributors might be the same. 

Yeah, it's not very old, @gziolo proposed this a couple months ago, It makes sense and simplify backports and such. And yes, we definitely need to document this in the folder structure page or more. 👍

File: lib/full-site-editing/edit-site-page.php (hunk 2 above)
Comment (youknowriad) [reply lib/full-site-editing/edit-site-page.php]: There's a failing end2end test, if I had to guess, I think it's related to the change here, meaning Core is probably also broken and missing a backport :)

**Comment by gziolo** [comment]
> > To be honest, I didn't know this until recently. I wonder if other contributors might be the same.
> 
> Yeah, it's not very old, @gziolo proposed this a couple months ago, It makes sense and simplify backports and such. And yes, we definitely need to document this in the folder structure page or more. 👍

Related issue: https://github.com/WordPress/gutenberg/issues/33810. We should formalize it once we have the structure in place. I know that @noisysocks started exploring those ideas a few months ago in the https://github.com/WordPress/gutenberg/tree/update/refactor-php branch, but I guess it was too much work for a single contributor.

**Comment by youknowriad** [comment]
Yes, we can't change everything in one step, doing it file by file seems a good compromise while also checking the impact compared to Core. For instance this just helped uncover an unbackported change dd377ed

File: lib/full-site-editing/edit-site-page.php (hunk 2 above)
Comment (youknowriad) [reply lib/full-site-editing/edit-site-page.php]: Ok so this should be solved with dd377ed and that commit need to be backported in Core properly.

**Comment by youknowriad** [comment]
I could use a ✅ here to continue iteration on these kind of changes.

File: lib/full-site-editing/template-loader.php (hunk 3 above)
Comment (ockham) [reply lib/full-site-editing/template-loader.php:19]: Yep! I see you're addressing that in #37149 👍

**Review by ockham — APPROVED** [review]
Thanks Riad! ✅

**Comment by noisysocks** [comment]
Agree with all of this 👍 thanks for tackling it @youknowriad.

A `README.md` in `lib` would be good to add. It could contain some best practices too. For example:

- It's easier to backport things when PHP is grouped together by feature (`full-site-editing.php`, `widgets.php`) instead of by functionality (`blocks.php`, `functions.php`).
- It's easier to backport things when the call to `add_filter`, `add_action` is right next to the function definition.
- It's easier to backport things when PHP doc comments are used to describe where in Core the functionality is supposed to land. (Or, even better, there's a link to a Trac ticket.)

**Pull request merged**